spec/lib/task-scheduler-spec.js

```
@@ -253,6 +253,77 @@ describe('Task Scheduler', function() {
+            var progressData = {
+                progress: {
+                    value: null,
+                    maximum: null,
+                    description: 'Task "' + taskFriendlyName + '" finished',
+                },
+                taskProgress: {
+                    taskId: data.task.instanceId,
+                    taskName: taskFriendlyName,
+                    progress: {
+                        value: 100,
+                        maximum: 100,
+                        description: "Task finished"
+                    }
+                }
+            };
+
+            return taskScheduler.publishTaskFinishedProgressEvent(task)
+            .then(function() {
+                expect(taskMessenger.publishProgressEvent).to.have.been.calledOnce;
+                expect(taskMessenger.publishProgressEvent)
+                    .to.have.been.calledWith(data.graphId, progressData);
+            });
+        });
+
+        it('GraphFinishedProgressEvent should publish with taskMessenger', function() {
+            this.sandbox.stub(taskMessenger, 'publishProgressEvent').resolves();
+            var graph = {
+                instanceId: 'testgraphid',
+                _status: 'succeeded',
+                node: 'nodeId',
+                definition: {
+                    friendlyName: "test graph name"
+                }
+            };
+
+            var progressData = {
+                progress: {
+                    maximum: null,
+                    value: 0,
+                    description: 'Graph "' + graph.definition.friendlyName + '" finished'
+                }
+            };
+
+            return taskScheduler.publishGraphFinishedProgressEvent(graph)
+            .then(function() {
+                expect(taskMessenger.publishProgressEvent).to.have.been.calledOnce;
+                expect(taskMessenger.publishProgressEvent)
+                    .to.have.been.calledWith(graph.instanceId, progressData);
+            });
+        });
+
```

There is a special feature for publish task/graph progress event is that, it will swallow the error if the publishing itself has some error, so that the event publishing will not impact the existing logic. So here need a test case to check the error is really being swallowed.